### ISSUE TITLE
edgeos_config command diff only considers router's active config

### ISSUE THREAD

**Issue by F1rst-Unicorn** [ansible/ansible#40437]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME

edgeos_config

##### ANSIBLE VERSION

```
ansible 2.5.3
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/jan/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.6/site-packages/ansible
  executable location = /bin/ansible
  python version = 3.6.5 (default, May 11 2018, 04:00:52) [GCC 8.1.0]

```

##### CONFIGURATION

```
CACHE_PLUGIN(/home/jan/Programme/Ansible/Schumacherweg/ansible.cfg) = jsonfile
CACHE_PLUGIN_CONNECTION(/home/jan/Programme/Ansible/Schumacherweg/ansible.cfg) = cache
CACHE_PLUGIN_TIMEOUT(/home/jan/Programme/Ansible/Schumacherweg/ansible.cfg) = 86400
DEFAULT_GATHERING(/home/jan/Programme/Ansible/Schumacherweg/ansible.cfg) = smart
DEFAULT_HOST_LIST(/home/jan/Programme/Ansible/Schumacherweg/ansible.cfg) = ['/home/jan/P>
DEFAULT_VAULT_PASSWORD_FILE(/home/jan/Programme/Ansible/Schumacherweg/ansible.cfg) = /ho>
RETRY_FILES_ENABLED(/home/jan/Programme/Ansible/Schumacherweg/ansible.cfg) = False
```

##### OS / ENVIRONMENT

Running on Arch Linux, managing EdgeOS v1.10.1 router 

##### SUMMARY

When comparing the active config with the desired commands, effects of previously scheduled desired commands is not taken into account.

##### STEPS TO REPRODUCE

Make sure no dhcp-server is present. 
Create a command chain to set up some config, starting from a clean state. Example set-style config:

```yaml
delete service dhcp-server
set service dhcp-server shared-network-name test
# ... possibly more instructions
```
Roll out with
```yaml
- edgeos_config:
    src: config-set.j2
```
Then roll out the playbook for a second time. 

##### EXPECTED RESULTS

When rolling out the playbook for the second time, no changes are applied to the system and no services are restarted. 

##### ACTUAL RESULTS

When rolling out for the second time, only the `delete` command is executed, leaving the router without any configured dhcp-server. 

-vvv output of the first run:
```yaml
changed: [ubnt01] => {
    "backup_path": "/home/jan/Programme/Ansible/Schumacherweg/roles/edgeos_dhcp_server/backup/ubnt01_config.2018-05-19@15:34:58",                                                 
    "changed": true,
    "commands": [
        "set service dhcp-server shared-network-name test"                                               
    ],
    "filtered": [],
    "invocation": {
        "module_args": {
            "backup": false,
            "comment": "configured by edgeos_config",
            "config": null,
            "lines": null,
            "match": "line",
            "save": false,
            "src": "delete service dhcp-server\nset service dhcp-server shared-network-name test"
        }
    }
}

```

-vvv output of the second run: 
```yaml
changed: [ubnt01] => {
    "backup_path": "/home/jan/Programme/Ansible/Schumacherweg/roles/edgeos_dhcp_server/backup/ubnt01_config.2018-05-19@15:34:09",                                                 
    "changed": true,
    "commands": [
        "delete service dhcp-server"
    ],
    "filtered": [],
    "invocation": {
        "module_args": {
            "backup": false,
            "comment": "configured by edgeos_config",
            "config": null,
            "lines": null,
            "match": "line",
            "save": false,
            "src": "delete service dhcp-server\nset service dhcp-server shared-network-name test"                        
        }
    }
}
```

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/network/edgeos/edgeos_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/network/edgeos/edgeos_config.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @qalthos @samdoran
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by samdoran** [comment]
Thank you for reporting this.

I am able to duplicate this and have determined why it is happening. Working on a fix, though it's a bit difficult.

**Comment by samdoran** [comment]
Can you test PR #40666 to see if that resolves the issue?

**Comment by F1rst-Unicorn** [comment]
I can confirm that the router config is not changed any more after running the task. No services are restarted and no commits are made. 

As a cosmetic issue ansible could also recognise nothing has changed and set the `changed` flag appropriately. At the moment the task is always reported as changed. But this might be out of scope here.

Thank you for your work and the fast response time!

**Comment by samdoran** [comment]
@F1rst-Unicorn Thanks for confirming this works.

Regarding the `changed flag`, that's tricky. I debated this when I was debugging and left it alone for this fix.

What's actually happening after this fix is, if a `delete` command is found in the desired config and a corresponding `set` command also exists in the desired config _as well as_ the running config, both commands will be run. So technically changes are being made to the router each time you run this task. Accurately determining whether or not a net change in the configuration was made would require a rather complex comparison.

The reason this gets tricky is say I have the following running config:

```
set service dhcp-server shared-network-name inside
set service dhcp-server shared-network-name guest
set service dhcp-server shared-network-name dmz
```

Then I run a playbook with the following config:
```
delete service dhcp-server
set service dhcp-server shared-network-name inside
```

The first run of the play will result in only the `inside` DHCP server existing, and subsequent runs will reissue the same `delete` and `set` commands, keeping the device state consistent with what is specified in the task.

We would have to know all the commands that the `delete` command removed, then compare those to the commands that will be set along with the commands that exist in the running config in order to determine if a net change was made. This is pretty complex and may not behave consistently given the complex nature of all the possible commands a device can accept.

It may very well be the case that adding this extra comparison is doable, but my brain still hurts from several days of trying to fix this issue. I may revisit this at a later time.

Hopefully I've explained the complexity of making this determination well. I'm very much open to input if you have any suggestions on how to reliably compare net changes. Thanks again for your feedback.
